src/System.Linq.Expressions/src/System/Linq/Expressions/StackGuard.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Linq.Expressions
+{
+    internal sealed class StackGuard
+    {
+        private const int StackProbingThreshold = 100;
+        private const int MaxExecutionStackCount = 1024;
+
+        private int _recursionDepth;
+        private int _executionStackCount;
+
+        public bool TryEnterOnCurrentStack()
+        {
+            if (++_recursionDepth >= StackProbingThreshold)
```

Comment:
Just ran some micro-benchmark on the Windows build; there's no measurable difference as far as I can tell. In fact, without it the measurement exhibit less variance and show marginally *better* numbers.

The initial concern brought up by @VSadov was portability of this function (and whether it's cheap on all platforms) but from looking at the code (and confirming with @ericeil) it seems it will merely do some arithmetic in a platform-independent manner.

I'm inclined to take out this additional complexity and am assuming the 99% case of running atop of user code will not cause `EnsureSufficientExecutionStack` to throw upon the call for the first iteration.